packages/block-editor/src/components/provider/index.js

```diff
@@ -6,6 +6,7 @@ import { last, noop } from 'lodash';
 /**
```

Comment:
I've been thinking a little bit about the Provider and controlled InnerBlocks and for me, both should do the exact same thing when it comes to "syncing" and calling onChange/onInput, so I have a few remarks/questions here

 - Why don't we have onInput/onChange on inner blocks, I believe the undo/redo for non-persistent changes should be broken now inside inner blocks (I mean if you type on a RichText inside a template part, we're probably creating an undo level for each character right?)
 - The syncing logic is becoming very complex and we need to test it properly
 - Can we extract the syncing logic into a React hook or something like that and reuse it in both components (InnerBlocks and Provider), an API like that maybe:

```
   useControlledBlockEditor( rootClientId, onInput, onChange )
```

Reply:
Also, this can be a separate PR too.

Reply:
>I've been thinking a little bit about the Provider and controlled InnerBlocks and for me, both should do the exact same thing when it comes to "syncing" and calling onChange/onInput,

Yeah, they do.

>Why don't we have onInput/onChange on inner blocks, I believe the undo/redo for non-persistent changes should be broken now inside inner blocks (I mean if you type on a RichText inside a template part, we're probably creating an undo level for each character right?)

No, inner blocks can still use the `onChange`/`onInput` logic.

>The syncing logic is becoming very complex and we need to test it properly

Agreed.

>Can we extract the syncing logic into a React hook or something like that and reuse it in both components (InnerBlocks and Provider), an API like that maybe:

It would be cleaner as a function because right now the block editor provider uses a global listener, while inner blocks handle this in a lifecycle event.

Reply:
> It would be cleaner as a function because right now the block editor provider uses a global listener, while inner blocks handle this in a lifecycle event.

Why can't we rely on the same approach, a hook or something else is not important but I'd like for these two components to share this logic somehow.

Reply:
I think sharing logic would be helpful too, I'll look into that. 

> The syncing logic is becoming very complex and we need to test it properly

+1. Sharing logic would help make testing easier too. The hook you suggested is actually very similar to the existing `useEntityBlockEditor`. I wonder if the logic should go there? At the same time, that hook is not used for edit-post so we might start touching more places.

> Why don't we have onInput/onChange on inner blocks

I think theoretically, `onInput` should still be called if it exists and if the blocks change. It's hard to test undo, since we haven't yet integrated undo shortcuts or buttons into the editor 😁

Reply:
I definitely think this would be useful for a follow-up, unless you think otherwise :)